opendevin/runtime/server/runtime.py

```diff
@@ -126,63 +126,76 @@ async def copy_to(self, host_src: str, sandbox_dest: str, recursive: bool = Fals
     async def run(self, action: CmdRunAction) -> Observation:
         return self._run_command(action.command)
 
+    def restart_kernel(self) -> str:
+        if self.config.default_agent in ['CodeActAgent', 'CodeActSWEAgent']:
+            kernel_init_code = 'from agentskills import *'
+        else:
+            return ''
+
+        restart_kernel_code = (
+            'import IPython\nIPython.Application.instance().kernel.do_shutdown(True)'
+        )
+        self._run_command(
+            (
+                "cat > /tmp/opendevin_jupyter_temp.py <<'EOL'\n"
+                f'{restart_kernel_code}\n'
+                'EOL'
+            )
+        )
+        obs = self._run_command('cat /tmp/opendevin_jupyter_temp.py | execute_cli')
+        output = obs.content
+        if "{'status': 'ok', 'restart': True}" != output.strip():
+            print(output)
```

Comment:
It does not happen yet.
```suggestion
            logger.error(f'Failed to restart the kernel: {output}')
```